File: hw1/hw1/Parser.cs
using System;

namespace hw1
{
    public static class Parser
    {
        public const int AllCorrect = 0;
        public const int WrongArgsLength = 1;
        public const int WrongArgFormat = 2;
        public const int WrongOperation = 3;

        private static bool CheckArgsLength(string[] args)
        {
            if (args.Length == 3)
            {
                return true;
            }

            Console.WriteLine($"The program requires 3 CLI arguments but {args.Length} provided");
            return false;
        }

        public static int TryParseArgs(string[] args, out int val1, out Operation operation, out int val2)
        {
            operation = default;
            val1 = default;
            val2 = default;

            if (!CheckArgsLength(args))
            {
                return WrongArgsLength;
            }

            if (!TryParseOperation(args[1], out operation))
            {
                return WrongOperation;
            }

            if (!TryParseValue(args[0], out val1) || !TryParseValue(args[2], out val2))
            {
                return WrongArgFormat;
            }

            return AllCorrect;
        }

        private static bool TryParseOperation(string arg, out Operation operation)
        {
            operation = arg switch
            {
                "+" => Operation.Plus,
                "-" => Operation.Minus,
                "*" => Operation.Multiply,
                "/" => Operation.Divide,
                _ => Operation.IncorrectOperation
            };
            var isOperation = operation is not Operation.IncorrectOperation;
            if (!isOperation)
            {
                Console.WriteLine($"Unsupported operation received: {arg}");
            }

            return isOperation;
        }

        private static bool TryParseValue(string arg, out int val)
        {
            var isVal = int.TryParse(arg, out val);
            if (!isVal)
            {
                Console.WriteLine($"Value is not int: {arg}");
            }

            return isVal;
        }
    }
}

File: hw4/hw4/Parser.fs
module hw4.Parser

let checkArgsLength (args:string[]) = args.Length = 3

let parseOperation (s:string) =
    match s with
    | "+" -> Operation.Plus
    | "-" -> Operation.Minus
    | "*" -> Operation.Multiply
    | "/" -> Operation.Divide
    | _ -> raise (Exceptions.WrongOperation $"Invalid operation: %s{s}")

let parseArgs (args:string[]) =
    if checkArgsLength args <> true then
        raise (Exceptions.WrongArgsLength $"3 arguments are required, but %i{args.Length} provided")

    let val1 =
        try
            args.[0] |> double
        with
            | _ -> raise (Exceptions.WrongArgFormat $"Wrong argument format: %s{args.[0]}")
    let val2 =
        try
            args.[2] |> double
        with
            | _ -> raise (Exceptions.WrongArgFormat $"Wrong argument format: %s{args.[2]}")

    let op = parseOperation args.[1]
    val1, op, val2